File: framework/test2/struct/lograte/lograte.py
#!/usr/bin/env python3
from collections import Counter, defaultdict, deque
import heapq
import json
import logging
import math
import random
import time
from typing import *
import sys

from soda.leetcode.bitree import *
from soda.leetcode.graph import *
from soda.leetcode.linklist import *
from soda.unittest.util import init_logging

logger = logging.getLogger(__name__)

# step [1]: implement class Solution
# class Solution: pass
class Logger:
    def __init__(self):
        self.msgMap = {}
        self.Limit = 10
        self.lastTs = -self.Limit

    def shouldPrintMessage(self, timestamp: int, message: str) -> bool:
        T = self.lastTs
        self.lastTs = timestamp
        if timestamp - T >= self.Limit:
            self.msgMap = {}
            self.msgMap[message] = timestamp
            return True
        if timestamp - self.msgMap.get(message, timestamp - self.Limit) < self.Limit:
            return False
        self.msgMap[message] = timestamp
        return True

if __name__ == '__main__':
    init_logging()
    from soda.unittest import Validators
    from soda.unittest.work import TestWork

    # step [2]: setup function
    # Attention! FUNCTION must use type hint, including arguments and return type
    # work = TestWork(Solution().FUNCTION)
    # OR use struct tester
    work = TestWork.forStruct(Logger)

    # step [3]: setup other options
    # work.validator = (e,r) => bool
    work.compareSerial = True
    work.run()

File: test-case/struct/0359/lograte.rb
# step [1]: implement solution function
class Logger
  def initialize()
    @msgMap = {}
    @limit = 10
    @lastTs = -@limit
  end

=begin
  :type timestamp: Integer
  :type message: String
  :rtype: Boolean
=end
  def should_print_message(timestamp, message)
    _T = @lastTs
    @lastTs = timestamp
    if timestamp - _T >= @limit
      @msgMap = {}
      @msgMap[message] = timestamp
      return true
    end
    if timestamp - @msgMap.fetch(message, timestamp - @limit) < @limit
      return false
    end
    @msgMap[message] = timestamp
    true
  end

end

if __FILE__ == $0
  require 'soda/unittest/work'
  ns = Soda::Unittest
  # step [2]: setup function/return/arguments
  # work = ns::TestWork.create(method(:my_solution))
  work = ns::TestWork.for_struct(Logger)
  # work.validator = -> (e, r) { e is equal to r }
  work.compare_serial = true
  puts work.run(ARGF.read)
end

